quickbench.1
.\" Automatically generated by Pandoc 3.6.4
.\"
.TH "quickbench" "1" "Apr 2025" "quickbench 1.1"
.SH NAME
quickbench \- quick & easy benchmarking of command\-line programs
.SH SYNOPSIS
\f[CR]quickbench [options] [CMD...]\f[R]
.SH DESCRIPTION
quickbench runs some test commands, possibly with different executables,
once or more and shows their best execution times in tabular format.
.PP
It produces very simple output (elapsed seconds), as quickly as possible
(running commands just once by default), and tabulates results from
multiple executables.
I find it useful for quick and dirty, exploratory, and comparative
.PD 0
.P
.PD
measurements that you (and others) can understand at a glance.
.PP
Commands are specified as one or more quote\-enclosed arguments, and/or
one per line in CMDSFILE; or read from a default file [./bench.sh].
.PP
With \-w, commands\[cq] first words are replaced with a new executable
(or multiple comma\-separated executables, showing times for all).
.SH OPTIONS
.TP
\f[CR]\-f, \-\-file CMDSFILE\f[R]
file containing commands, one per line (\- for stdin)
.TP
\f[CR]\-w, \-\-with EXE[,...]\f[R]
replace first word of commands with these executables
.TP
\f[CR]\-n, \-\-iterations=N\f[R]
run each test this many times [default: 1]
.TP
\f[CR]\-N, \-\-cycles=N\f[R]
run the whole suite this many times [default: 1]
.TP
\f[CR]\-p, \-\-precision=N\f[R]
show times with this many decimal places [default: 2]
.TP
\f[CR]\-v, \-\-verbose\f[R]
show commands being run
.TP
\f[CR]\-V, \-\-more\-verbose\f[R]
show command output
.TP
\f[CR]\-\-debug\f[R]
show debug output for this program
.TP
\f[CR]\-h, \-\-help\f[R]
show this help
.SH EXAMPLES
.IP
.EX
$ quickbench \[aq]sleep 1\[aq]
Running 1 tests 1 times at 2016\-10\-16 23:06:48.058578 UTC:

Best times:
+\-\-\-\-\-\-\-\-\-++\-\-\-\-\-\-+
|         ||      |
+=========++======+
| sleep 1 || 1.01 |
+\-\-\-\-\-\-\-\-\-++\-\-\-\-\-\-+
.EE
.IP
.EX
$ echo \[aq]echo 3 * 1000000\[aq] > bench.sh
$ quickbench \-w echo,expr \-p5 \-n100 \-N2
Running 1 tests 100 times with 2 executables at 2016\-10\-16 23:57:34.387764 UTC:

Best times 1:
+\-\-\-\-\-\-\-\-\-\-\-\-\-++\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-+
|             ||    echo |    expr |
+=============++=========+=========+
| 3 * 1000000 || 0.00112 | 0.00135 |
+\-\-\-\-\-\-\-\-\-\-\-\-\-++\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-+

Best times 2:
+\-\-\-\-\-\-\-\-\-\-\-\-\-++\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-+
|             ||    echo |    expr |
+=============++=========+=========+
| 3 * 1000000 || 0.00111 | 0.00136 |
+\-\-\-\-\-\-\-\-\-\-\-\-\-++\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-+
.EE
.SH FILES
quickbench looks for tests in \f[CR]./bench.sh\f[R] if no command
arguments or \f[CR]\-\-file\f[R] option are provided.
.SH LIMITATIONS
quickbench tests executable files on disk only, not shell builtins or
aliases.
.PP
Options must precede arguments on the command line.
.PP
\f[CR]\-V\f[R] will become \f[CR]\-vv\f[R] when implementing that
becomes easier.
.SH BUG REPORTS
https://github.com/simonmichael/quickbench/issues
.SH SEE ALSO
Home page: https://github.com/simonmichael/quickbench
.PP
time(1)
.SH COPYRIGHT
Copyright (C) 2008\-2025 Simon Michael.
Released under GNU GPL v3+
.SH AUTHORS
Simon Michael \c
.MT simon@joyful.com
.ME \c.


CHANGES
1.0.1 2021-06-15

- update README, build instructions, & stack.yaml file.

1.0 2016-10-21

- extract tools/simplebench from the hledger repo

- stop using error, make safe for reuse

- more robust command execution, signal handling, output, error reporting

- switch to docopt, options must precede args for now

- new UI: full commands are specified by arguments/file/stdin,
  alternate executables can be specified with -w, flags cleanup

- default file for test commands is ./bench.sh

- -N/--cycles repeats the whole test suite

- man page


src/QuickBench.hs
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeOperators #-}

module QuickBench
-- (
--   defaultMain
-- )
where

-- import Debug.Trace
import Control.Exception
import Control.Monad
import Data.Char (isSpace)
import Data.List
import Data.List.Split (splitOn)
import Data.Maybe
import Data.Time.Clock
import Data.Time.Format
import Data.Time.LocalTime
import GHC.IO.Exception (IOErrorType(..))
import Safe
import System.Console.Docopt
import System.Directory
import System.Environment
import System.Exit
import System.IO
import System.IO.Error (mkIOError)
import System.Process
import Text.Megaparsec (ParsecT, Stream (Token), between, many, noneOf, runParser, satisfy, sepBy, takeWhile1P, (<|>))
import Text.Megaparsec.Char (char)
import Text.Show.Pretty
import Text.Printf
import Text.Tabular
import qualified Text.Tabular.AsciiArt as TA


-- Command line help, parsed by docopt to generate the CLI.
-- This is fragile and easy to break; always test after any changes.
-- See also: https://github.com/docopt/docopt.hs?tab=readme-ov-file#help-text-format
-- https://hackage.haskell.org/package/docopt-0.7.0.8/docs/System-Console-Docopt.html
-- Keep version & doc synced: here, quickbench.cabal, quickbench.1.md, README.md
docoptpatterns :: Docopt
docoptpatterns = [docopt|
------------------------------------------------------------------------------
quickbench 1.1 - run commands and show how long they took.

Usage:
  quickbench [options] [CMD...]

Each command's first word should be an executable (not a shell builtin).
Multi-word commands must be enclosed in quotes.
Commands can also be read from a file specified with -f.
If no commands are provided, it looks for them in ./bench.sh.
Any quickbench options should be written first, before the commands.

Options:
  -f, --file CMDFILE    run commands from this file (- for stdin)
  -w, --with EXE[,...]  run commands with first word replaced by each EXE
  -n, --iterations=N    run each command this many times [default: 1]
  -N, --cycles=N        run the whole suite this many times [default: 1]
  -p, --precision=N     show times with this many decimal places [default: 2]
  -v, --verbose         show the commands being run
  -V, --more-verbose    show the commands' output
      --debug           show this program's debug output
  -h, --help            show this help

|]

defaultFile :: FilePath
defaultFile = "bench.sh"

data Opts = Opts {
   file        :: Maybe FilePath
  ,executables :: [String]
  ,iterations  :: Int
  ,cycles      :: Int
  ,precision   :: Int
  ,verbose     :: Bool
  ,moreVerbose :: Bool
  ,debug       :: Bool
  ,help        :: Bool
  ,clicmds     :: [String]
} deriving (Show)

getOpts :: IO Opts
getOpts = do
  cliargs <- getArgs
  case parseArgs docoptpatterns cliargs of
    Left e -> putStrLn "Could not parse command line:" >> print e >> exitFailure
    Right dopts -> do
      let
        flag f = dopts `isPresent` longOption f
        opt f  = dopts `getArg`    longOption f
        readint s =
          case readMay s of
            Just a  -> return a
            Nothing -> fail $ "could not read " ++ show s ++ " as an integer"
        (lateflags, args) = partition ("-" `isPrefixOf`) $ dopts `getAllArgs` (argument "CMD")
      -- These will have a value because of their [default: ...] above.
      -- Use fromJust to show an error if the default is accidentally removed.
      iterations' <- readint $ fromJust $ opt "iterations"
      cycles'     <- readint $ fromJust $ opt "cycles"
      precision'  <- readint $ fromJust $ opt "precision"
      let
        opts = Opts {
          file         = opt "file"
          ,executables = maybe [] (splitOn ",") $ opt "with"
          ,iterations  = iterations'
          ,cycles      = cycles'
          ,precision   = precision'
          ,verbose     = flag "verbose"
          ,moreVerbose = flag "more-verbose"
          ,debug       = flag "debug"
          ,help        = flag "help"
          ,clicmds     = args
          }

      when (debug opts || "--debug" `elem` lateflags) $ do
        err $ "docopts: " ++ ppShow dopts ++ "\n" ++ ppShow opts ++ "\n"

      when (help opts) $ putStrLn (usage docoptpatterns) >> exitSuccess

      -- try to report some errors docopts misses
      case (lateflags, args) of
        -- unknown option
        -- quickbench a -fk       user error (unknown option: "fk")
        -- (f:_,[]) | not $ elem f [
        --    "file","f"
        --   ,"with","w"
        --   ,"iterations","n"
        --   ,"cycles","N"
        --   ,"precision","p"
        --   ,"verbose","v"
        --   ,"more-verbose","V"
        --   ] -> fail $ printf "unknown option: %s" (show f)
        -- option value missing
        (f:_,[]) -> fail $ printf "option %s needs a value or is unknown" (show f)
        -- option following arguments
        (f:_,a:_) -> fail $ printf "option %s should appear before argument %s or is unknown" (show f) (show a)
        _ -> return ()
      return opts

-- | Run the quickbench program, returning an error message if there was a problem.
defaultMain :: IO (Maybe String)
defaultMain =
  (runSuite >> return Nothing)
    `catch` \(e :: SomeException) -> return $
      if fromException e == Just ExitSuccess
      then Nothing
      else Just $ show e
  where
    runSuite = do
      opts <- getOpts
      let istest s = not (null s' || ("#" `isPrefixOf` s')) where s' = strip s
      filecmds <-
        (filter istest . lines) <$>
        (case (file opts, clicmds opts) of
          (Just "-", _) -> getContents
          (Just f, _)   -> readFile f
          (Nothing, []) -> doesFileExist defaultFile >>=
            \yes -> if yes then readFile defaultFile else return ""
          (Nothing, _)  -> return [])
      let cmds = filecmds ++ clicmds opts
      when (null cmds) $ do
        out opts "No test commands found; provide some as arguments, with -f, or in ./bench.sh\n"
        exitSuccess
      now <- getCurrentZonedTime
      out opts $ printf "Running %s%d tests best of %d times%s at %s:\n"
        (case cycles opts of
          0 -> ""
          n -> show n <> " cycles of "
          )
        (length cmds)
        (iterations opts)
        (case executables opts of
          [] -> ""
          es -> printf " with %d executables" (length es))
        (formatTime defaultTimeLocale "%Y-%m-%d %T %Z" now)
      let
        exes = case executables opts of
          [] -> [""]
          es -> es
      hSetBuffering stdout NoBuffering
      forM_ [1..cycles opts] $ \cyc -> do
        results <- mapM (runTestWithExes opts exes) cmds
        printSummary opts cmds exes cyc results

getCurrentZonedTime :: IO ZonedTime
getCurrentZonedTime = do
  t <- getCurrentTime
  tz <- getCurrentTimeZone
  return $ utcToZonedTime tz t

runTestWithExes :: Opts -> [String] -> String -> IO [[Float]]
runTestWithExes opts exes cmd = mapM (runTestWithExe opts cmd) exes

runTestWithExe :: Opts -> String -> String -> IO [Float]
runTestWithExe opts cmd exe = mapM (runTestOnce opts cmd exe) [1..iterations opts]

runTestOnce :: Opts -> String -> String -> Int -> IO Float
runTestOnce opts cmd exe iteration = do
  let (cmd',exe',args) = replaceExecutable exe cmd
  when (not $ null exe) $ dbg opts $ "replaced executable with " <> show exe
  outv opts (show iteration ++ ": " ++ cmd' ++ "\n")
  t <- time opts exe' args
  outv opts $ printf "\t[%ss]\n" (showtime opts t)
  return t

-- XXX confusing
-- | Replace a command line's command (first word) with the given executable.
-- Returns the new command line, its command, and its arguments.
-- If the command line was empty, the executable is used as the new command line.
-- If the executable was empty, the command line remains unchanged.
-- (And if both were empty, return empty strings.)
-- If the executable consists of multiple words, the first becomes the new
-- command, and the rest are treated as its initial arguments.
replaceExecutable :: String -> String -> (String,String,[String])
replaceExecutable exestring cmdline = (unwords argv, headDef "" argv, tailSafe argv)
  where
    exeWords = words' exestring
    cmdWords = words' cmdline
    argv = case exeWords of
      [] -> cmdWords
      _  -> exeWords ++ drop 1 cmdWords

time :: Opts -> String -> [String] -> IO Float
time opts exe args = do
  dbg opts $ printf "running: %s\n" (show (exe,args))
  t1 <- getCurrentTime
  maybeOutput <- if (moreVerbose opts)
    then Just <$> readCreateProcess (proc exe args) ""
    else const Nothing <$> callProcessIgnoreOutput exe args
  t2 <- getCurrentTime
  case maybeOutput of
    Just o -> when (not $ null o) $ outvv opts $ (if verbose opts then "\n" else "") ++ o
    Nothing -> return ()
  return $ realToFrac $ diffUTCTime t2 t1

callProcessIgnoreOutput :: FilePath -> [String] -> IO ()
callProcessIgnoreOutput cmd args =
  withBinaryFile "/dev/null" WriteMode $ \devNull ->
    withCreateProcess (proc cmd args){std_out = UseHandle devNull} $ \_ _ _ ph -> do
      exit_code <- waitForProcess ph
      case exit_code of
        ExitSuccess   -> return ()
        ExitFailure r -> processFailedException "callProcessIgnoreOutput" cmd args r

-- Copy/paste from "process" System.Process.
processFailedException :: String -> String -> [String] -> Int -> IO a
processFailedException fun cmd args exit_code =
      ioError (mkIOError OtherError (fun ++ ": " ++ cmd ++
                                     concatMap ((' ':) . show) args ++
                                     " (exit " ++ show exit_code ++ ")")
                                 Nothing Nothing)

printSummary :: Opts -> [String] -> [String] -> Int -> [[[Float]]] -> IO ()
printSummary opts cmds exes cyc results = do
  out opts $ printf "\nBest of %d times%s:\n" (iterations opts) (if cycles opts > 1 then ", cycle "++show cyc else "")
  let t = maketable opts cmds' exes results
  out opts $ TA.render id id id t
  -- let outname = "benchresults"
  -- writeFile (outname <.> "txt") $ TA.render id id id t
  -- writeFile (outname <.> "html") $ renderHtml $ TH.css TH.defaultCss +++ TH.render stringToHtml stringToHtml stringToHtml t
  where
    cmds' =
      case executables opts of
        []  -> cmds
        [e] -> [c | (c,_,_) <- map (replaceExecutable e) cmds]
        _   -> map (unwords . drop 1 . words) cmds

maketable :: Opts -> [String] -> [String] -> [[[Float]]] -> Table String String String
maketable opts rownames colnames results = Table rowhdrs colhdrs rows
 where
  rowhdrs = Group NoLine $ map Header $ padright rownames
  colhdrs = Group SingleLine $ map Header colnames
  rows = map (map (showtime opts . minimum)) results
  padright ss = map (printf (printf "%%-%ds" w)) ss
      where w = maximum $ map length ss

---------------------------------------
-- utils

-- IO

out :: Opts -> String -> IO ()
out _ = putStr

outv :: Opts -> String -> IO ()
outv opts s = when (verbose opts || moreVerbose opts) $ putStr s

outvv :: Opts -> String -> IO ()
outvv opts s = when (moreVerbose opts) $ putStr s

err :: String -> IO ()
err = hPutStr stderr

dbg :: Opts -> String -> IO ()
dbg opts s = when (debug opts) $ err s

showtime :: Opts -> (Float -> String)
showtime opts = printf $ "%." ++ show (precision opts) ++ "f"

-- Strings

-- | Remove leading and trailing whitespace.
strip :: String -> String
strip = lstrip . rstrip

-- | Remove leading whitespace.
lstrip :: String -> String
lstrip = dropWhile isSpace

-- | Remove trailing whitespace.
rstrip :: String -> String
rstrip = reverse . lstrip . reverse

-- XXX megaparsec dependency was added just for this
-- Quote-aware version of words - don't split on spaces which are inside quotes.
-- NB correctly handles "a'b" but not "''a''".
-- Partial, can raise an error if parsing fails.
words' :: String -> [String]
words' ""  = []
words' str = map stripquotes $ either errexit id $ runParser p "" str
 where
  errexit = errorWithoutStackTrace . ("parse error at " ++) . show
  p = (singleQuotedPattern <|> doubleQuotedPattern <|> patterns) `sepBy` skipNonNewlineSpaces1
  singleQuotedPattern = between (char '\'') (char '\'') (many $ noneOf "'")
  doubleQuotedPattern = between (char '"') (char '"') (many $ noneOf "\"")
  patterns = many (satisfy $ not . isSpace)
  skipNonNewlineSpaces1 :: (Stream s, Token s ~ Char) => ParsecT String s m ()
  skipNonNewlineSpaces1 = void $ takeWhile1P Nothing $ \c -> not (c == '\n') && isSpace c

  -- | Strip one matching pair of single or double quotes on the ends of a string.
  stripquotes :: String -> String
  stripquotes s = if isSingleQuoted s || isDoubleQuoted s then init $ tailErr s else s -- PARTIAL tailErr won't fail because isDoubleQuoted

  isSingleQuoted :: String -> Bool
  isSingleQuoted s@(_ : _ : _) = headErr s == '\'' && last s == '\'' -- PARTIAL headErr, last will succeed because of pattern
  isSingleQuoted _ = False

  isDoubleQuoted :: String -> Bool
  isDoubleQuoted s@(_ : _ : _) = headErr s == '"' && last s == '"' -- PARTIAL headErr, last will succeed because of pattern
  isDoubleQuoted _ = False
